[FLINK-12541][container][python] Add support for Python jobs in build script

dianfu opened the pull request:
## What is the purpose of the change

*This pull request adds support to build job specific docker image for Python Table API jobs.*

## Brief change log

  - *Improves the build script to support build a Python job specific docker image*

## Verifying this change

Verify this change manually.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

sunjincheng121 commented:
@flinkbot attention @tillrohrmann

tillrohrmann commented on flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java:
```diff
@@ -73,27 +76,31 @@
-		this(jobId, savepointRestoreSettings, programArguments, jobClassName, JarsOnClassPath.INSTANCE);
+			@Nullable String jobEntryPointName,
+			@Nullable String jobPythonArtifacts) {
+		this(jobId, savepointRestoreSettings, programArguments, jobEntryPointName, jobPythonArtifacts, JarsOnClassPath.INSTANCE);
```
Why do we need to touch the `ClassPathJobGraphRetriever` at all? This class should not need to know about Python. Otherwise we need to add special casing for all supported languages in the future. This does not seem right.

tillrohrmann commented on flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java:
```diff
@@ -110,19 +117,38 @@ public JobGraph retrieveJobGraph(Configuration configuration) throws FlinkExcept
 			throw new FlinkException("Could not load the provided entrypoint class.", e);
+		} catch (ClassNotFoundException e) {
+			if (!isPythonProgram && jobPythonArtifacts != null) {
+				return createPackagedProgram("org.apache.flink.python.client.PythonDriver", true);
```
This is overly complicated and error prone. Why not starting the cluster entrypoint with `o.a.f.python.client.PythonDriver` as `jobClassName`?

tillrohrmann commented on flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java:
```diff
@@ -110,19 +117,38 @@ public JobGraph retrieveJobGraph(Configuration configuration) throws FlinkExcept
+			args[2] = "pyfs";
+			args[3] = jobPythonArtifacts;
+			System.arraycopy(programArguments, 0, args, 4, programArguments.length);
+			return args;
```
Why do we need to do this command line argument magic? Just because `PythonDriver` expects the arguments to be passed in a special order? Why not making the `PythonDriver` more flexible.

tillrohrmann commented on flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfiguration.java:
```diff
@@ -69,7 +74,12 @@ JobID getJobId() {
+	}
+
+	@Nullable
+	String getJobPythonArtifacts() {
```
Why is this needed?

tillrohrmann commented on flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactory.java:
```diff
@@ -45,12 +45,29 @@
+		.desc("Entrypoint name of the job to run.")
+		.build();
+
+	private static final Option JOB_PYTHON_ARTIFACTS_OPTION = Option.builder("ja")
```
Why does this need to be handled by the `StandaloenJobClusterConfiguration` and cannot be parsed by the `PythonDriver` program?

tillrohrmann commented on flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPoint.java:
```diff
@@ -53,26 +53,31 @@
+	private final String jobEntryPointName;
+
+	@Nullable
+	private final String jobPythonArtifacts;
```
I think it is not a good idea to add special case logic into the `StandaloneJobClusterEntrypoint`. Why do you think this is needed?

tillrohrmann commented:
Why has this PR the same Flink issue assigned as #8532? Every PR should have it's own JIRA issue as stated in the [contribution guidelines](https://flink.apache.org/contribute-code.html).

sunjincheng121 commented:
Thanks for the PR @dianfu 
I think the suggestion from @tillrohrmann makes more sense to me. we should let the PR have its own JIRA.
furthermore,  the improvement of `pyfs` also can in a new PR (with a new JIRA).
What do you think?

dianfu commented:
@tillrohrmann Thanks a lot for your review. 
Your suggestion makes much sense to me. I have created a dedicated JIRA [FLINK-12788](https://issues.apache.org/jira/browse/FLINK-12788) for this PR. Regarding to the changes to `StandaloneJobClusterEntrypoint`, agree that there should not be special logic for Python. I will revert that part of changes.

@sunjincheng121 Thanks a lot for your review. I have created a separate JIRA [FLINK-12787](https://issues.apache.org/jira/browse/FLINK-12787) for the PythonDriver improving. Then we can focus this PR on the build script changes for Python jobs support. 

Will updated the PR later today.

dianfu commented:
@tillrohrmann @sunjincheng121 I have updated the PR and the changes are only related to the build script for Python jobs support. Looking forward to your feedback.

sunjincheng121 commented:
Thanks for the update @dianfu!

Currently, the PR looks pretty clean form my side. only one improvement I am not pretty sure that is build command option `--opt-jars`.  this way the opt jar does not need upload. but without this option, the user can build a fat jar with option jars. So this change is an improvement(not necessary change) So we may need some opinion from @tillrohrmann!
Otherwise the PR LGTM. I'll merge it when @tillrohrmann says ok!

Best,
Jincheng

sunjincheng121 commented:
I think we can remove the `--opt-jars` due to we @StephanEwen already bring up the discussion for put the table JAR into lib. detail can be found here:  http://apache-flink-mailing-list-archive.1008284.n3.nabble.com/DISCUSS-Putting-Table-API-jars-in-lib-by-default-td29492.html
What do you think?

dianfu commented:
@sunjincheng121 Make sense to me as `--opt-jars` is not a must have option. Considering that the table JAR will be put into lib by default in the future, removing this option is good for me as the table JAR is the motivation for this option. Have updated the PR. Looking forward to your response.

tillrohrmann commented on flink-container/kubernetes/README.md:
```diff
@@ -29,7 +29,7 @@ In non HA mode, you should first start the job cluster service:
 
 In order to deploy the job cluster entrypoint run:
 
-`FLINK_IMAGE_NAME=<IMAGE_NAME> FLINK_JOB=<JOB_NAME> FLINK_JOB_PARALLELISM=<PARALLELISM> envsubst < job-cluster-job.yaml.template | kubectl create -f -`
```
Why did we remove `FLINK_JOB` here?

dianfu replied:
I found that FLINK_JOB has been removed in commit [753e0c6](https://github.com/dianfu/flink/commit/753e0c618a30a19954761cd75fadba4a89836ae1) and just correct the documentation here. What do you think?

tillrohrmann replied:
Good catch. Now this makes sense.

tillrohrmann commented on flink-container/docker/README.md:
```diff
@@ -13,26 +13,28 @@ Install the most recent stable version of [Docker](https://docs.docker.com/insta
 
 Before building the image, one needs to build the user code jars for the job.
-Assume that the job jar is stored under `<PATH_TO_JOB_JAR>` 
+Assume that the job jar is stored under `<COMMA_SEPARATED_PATH_TO_JOB_ARTIFACTS>`
```
Seems not to fit. I guess the sentence needs to be changed to something like: `A Flink job can consist of multiple jars. In order to specify the required jars, they need to be passed to `--job-artifacts` of the build script. The individual paths are comma separated.

dianfu replied:
+1. Only one concern: what about changing `jars` to `artifacts`?

dianfu commented:
@tillrohrmann Thanks a lot for your comments. I have updated the PR and addressed one of your concerns. Regarding to FLINK_JOB, this option has been removed and the documentation has not been updated. Just correct the documentation here. Please let me know if you feel that change doesn't make sense or you prefer to address that in another PR.

knaufk commented:
FYI: This change broke `test_docker_embedded_job.sh` and `test_kubernetes_embedded_job.sh` as they were still using the `--job-jar` argument. I have included a `[hotfix]` in https://github.com/apache/flink/pull/874.

dianfu commented:
@knaufk Thanks a lot for the hotfix. The fix makes sense to me.
